DOC: stats: probability tutorial/transition guide (#22025)
scipy/scipy: doc/source/tutorial/stats/rv_infrastructure.md

data = stats.norm.rvs(size=100, loc=mu, scale=sigma)
stats.norm.nnlf((mu, sigma), data)
```

Now, simply compute the negative [log-likehood function](https://en.wikipedia.org/wiki/Likelihood_function) according to its mathematical definition.

```{code-cell} ipython3
X = stats.Normal(mu=mu, sigma=sigma)
-X.logpdf(data).sum()
```

#### Expected Values

+++

The `expect` method of the old infrastructure estimates a definite integral of an arbitrary function weighted by the PDF of the distribution. For instance, the fourth moment of the distribution is given by:

```{code-cell} ipython3
def f(x): return x**4
stats.norm.expect(f, lb=-np.inf, ub=np.inf)
```

This provides little added convencience over what the source code does: use `scipy.integrate.quad` to perform the integration numerically.

```{code-cell} ipython3
from scipy import integrate
def f(x): return x**4 * stats.norm.pdf(x)
integrate.quad(f, a=-np.inf, b=np.inf)  # integral estimate, estimate of the error
```

Of course, this can just as easily be done with the new infrastructure.

```{code-cell} ipython3
def f(x): return x**4 * X.pdf(x)
integrate.quad(f, a=-np.inf, b=np.inf)  # integral estimate, estimate of the error

> ```

Author: It wasn't removed. Users are welcome to use `expect` from the old interface. It hasn't been a priority for me to add to the new interface because be really can't do much to improve upon the existing numerical integration features (or those we have recently added). Faced with the choice between spending time recreating these interfaces vs adding value, I chose the latter. Also, sometimes users seem to expect more from `expect` than it can reasonably provide (e.g. gh-8928, gh-9259, gh-14347). If `expect` were not an option, perhaps users will learn to perform numerical integration with the general tools. At their simplest, the general tools are not any more complex to use, but they give you a lot more control when things don't "just work".

```{code-cell} ipython3
stats.weibull_min.nnlf((c, loc, scale), data)
```

The NLLF of the parameters estimated using `fit` are lower:

```{code-cell} ipython3
stats.weibull_min.nnlf(params, data)
```

Therefore, `fit` considers its job done, whether or not this satisfies the user's notions of a "good fit" or whether the parameters are within reasonable bounds.

At its simplest, then, it is not so hard to replicate what `fit` does using tools available in SciPy. As discussed above, the NLLF is given by:

```{code-cell} ipython3
def nllf(params):
    c, loc, scale = params
    X = Weibull(c=c) * scale + loc
    return -X.logpdf(data).sum()

nllf(params)
```

To perform the minimization, we can use `scipy.optimize.minimize`.

```{code-cell} ipython3
from scipy import optimize
eps = 1e-10  # numerical tolerance to avoid invalid parameters
lb = [eps, -10, eps]  # lower bound on `c`, `location`, and `scale`
ub = [10, np.min(data)-eps, 10]  # upper bound on `c`, `location`, and `scale`
x0 = [1, 1, 1]  # guess to get optimization started
bounds = optimize.Bounds(lb, ub)
res = optimize.minimize(nllf, x0, bounds=bounds)
res
```

> Although the value of `fun` (the NLLF) is not quite as low as with the solution provided by `weibull_min.fit`, the PDFs are not so different.

Author: From the top:

> I plan to add quite a bit more about fitting.

Maybe if I haden't done made all the improvements to the existing `fit` this would have been an easier sell. Alas.

> Although the value of `fun` (the NLLF) is not quite as low as with the solution provided by `weibull_min.fit`, the PDFs are not so different.

Author: I made some improvements. I would like to expand with examples of fitting with other objective functions and entirely different requirements (e.g. users have asked to fit a distribution not to *data* but such that specific statistics assume particular values), but I think it is much more convincing. I didn't even consciously choose this case to be problematic for MLE, but it happens to be a great example of the sort of pitfals I've seen in "bug" reports.

```{code-cell} ipython3
Cauchy = stats.make_distribution(stats.cauchy)
c = 4.72

X = Cauchy() + c  
Y = abs(X)  # compare to `foldcauchy`
Y.plot(t=('x', 0, 60));

x = np.linspace(0, 60, 300)
plt.plot(x, stats.foldcauchy(c=c).pdf(x), '--')
```

`scipy.stats.lognormal` describes the exponential of a normally distributed random variable. It is so-named because the logarithm of the resulting random variable is normally distributed. (In general, a "log" distribution is typically the distribution underlying the *exponential* of a random variable.)

```{code-cell} ipython3
u, s = 1, 0.5

X = stats.Normal()*s + u
Y = stats.exp(X)  # compare to `lognorm`
Y.plot(t=('x', 0, 9));

x = np.linspace(0, 9, 300)
plt.plot(x, stats.lognorm(s=s, scale=np.exp(u)).pdf(x), '--')
```

`scipy.stats.loggamma` describes the logarithm of of a Gamma-distributed random variable. Note that the more common name of this distribution would be [exp-gamma](https://reference.wolfram.com/language/ref/ExpGammaDistribution.html.en#:~:text=The%20term%20exp%2Dgamma%20is,of%20qualitatively%20similar%20probability%20distributions.) because the exponential of the RV is gamma-distributed.

```{code-cell} ipython3

> Gamma = stats.make_distribution(stats.gamma)

Reviewer: What's the rationale for having `X = stats.Normal()` but not `X = stats.Gamma()`?

Author: For the _first_ distribution with a new class, `Normal` seemed appropriate. It's pretty fundamental.

Personally, I've envisioned adding a few of the most fundamental distributions back with the new naming scheme because the performance, accuracy, and interface can be a bit better than if we rely on `make_distribution` for them. For instance, `Uniform` can be parametrized by the endpoints rather than location/scale, and new options for private methods can improve the speed and accuracy of the public ones. As for why that's not *all* done yet, well, it's mostly a matter of time. `make_distribution` barely made it into 1.15, and that seemed more important to me, so that's where I spent my time.

Reviewer: > As for why that's not all done yet, well, it's mostly a matter of time. make_distribution barely made it into 1.15, and that seemed more important to me, so that's where I spent my time.

Completely understandable. Might still be worth a thought to note that the new infrastructure will still receive updates & improvements in the . The text kinda reads like it's in its final state already.

Author: Hah, that would be nice. Please see if the (new) conclusion begins to address this point, and let me know if there are particular places where the language might be changed to avoid the appearance that we think this is all done.

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

> (rv_infrastructure)=
> ```{eval-rst}
> .. jupyterlite:: ../../_contents/hypothesis_bartlett.ipynb
>    :new_tab: True
> ```
> 
> # Random Variable Transition Guide

Author: ```suggestion
```{eval-rst}
.. jupyterlite:: ../../_contents/rv_infrastructure.ipynb
   :new_tab: True
```

(rv_infrastructure)=
# Random Variable Transition Guide
```

Both approaches are valid and have advantages in certain situations. For instance, `stats.norm.pdf(x)` appears more natural than `stats.norm().pdf(x)` for such a simple invocation or when methods are viewed as functions of distribution parameters rather than the usual arguments (e.g., likelihood). However, the former approach has a few inherent disadvantages; e.g., all of SciPy's 125 continuous distributions have to be instantiated at import time, distribution parameters must be validated every time a method is called, and documentation of methods must either a) be generated separately for every method of every distribution or b) omit the shape parameters that are unique for each distribution.
]
To address these and other shortcomings, [gh-15928](https://github.com/scipy/scipy/issues/15928) proposed a new, separate infrastructure based on the latter (random variable) approach. This transition guide documents how users of `rv_continuous` and `rv_continuous_frozen` can migrate to the new infrastructure.

> *Note: The new infrastructure may not yet be as convenient for some use cases, especially fitting distribution parameters to data. Users are welcome to continue to use the old infrastructure if it suits their needs; it is not deprecated at this time.*

Reviewer: 👍

+++

## Basics

+++

> In the new infrastructure, distributions families are classes named according to `CamelCase` conventions. They must be instantiated before use, with parameters passed as keyword-only arguments.

Reviewer: ```suggestion
In the new infrastructure, distributions families are classes named according to `CamelCase` conventions (note that not all distributions have a corresponding class yet; see `make_distribution` below). They must be instantiated before use, with parameters passed as keyword-only arguments.
```

Author: 👍 I put something like this a little here.

- `logcdf` (logarithm of cumulative distribution function)
- `entropy` (differential entropy)
- `median`
- `support`

+++

Others methods have new names, but are otherwise drop-in replacements.
- `sf` (survival function) $\rightarrow$ `ccdf` (complementary cumulative distribution function)
- `logsf` $\rightarrow$ `logccdf`
- `ppf` (percent point function) $\rightarrow$ `icdf` (inverse cumulative distribution function)
- `isf` (inverse survive function) $\rightarrow$ `iccdf` (inverse complementary cumulative distribution function)
- `std` $\rightarrow$ `standard_deviation`
- `var` $\rightarrow$ `variance`

+++

The new infrastructure has several *new* methods in the same vein as those above.

- `ilogcdf` (inverse of the logarithm of the cumulative distribution function)
- `ilogccdf` (inverse of the logarithm of the complementary cumulative distribution function)
- `logentropy` (logarithm of the entropy)
- `mode` (mode of the distribution)
- `skewness`
- `kurtosis` (*non-excess* kurtosis; see "Standardized Moments" below)

+++

And it has a new `plot` method for convenience

```{code-cell} ipython3
import matplotlib.pyplot as plt
X.plot()
plt.show()

> ```

Reviewer: The default plot title is a bit awkward `Normal(mu=np.float64(0.0), sigma=np.float64(1.0))`. We could add a title manually; in an case, this is not a requirement, just a comment.

Author: Actually if @tylerjereddy will allow a backport for this, this is a little regression in the rendering. In a few lines, @steppi could add `__str__` overrides very similar to the existing `__repr__` overrides that would fix this.

Reviewer: Since it looks doc-only, I'm in favour of doing that.

Reviewer 2: @mdhaber, would rather I push a commit with the `__str__` overrides here, or make a separate PR against `main`?

Author: Resolved in gh-22081

```{code-cell} ipython3
# params = stats.weibull_min.fit(data, loc=0)  # careful! this provides loc=0 as a *guess*
params = stats.weibull_min.fit(data, floc=0)
params
```

```{code-cell} ipython3
stats.weibull_min.nnlf(params, data)
```

By relying on a `fit` method to "just work", users are encouraged to neglect some of these important details. 

Instead, we suggest that it is almost as simple to fit distributions to data using the generic optimization tools provided by SciPy. However, this allows the user to perform the fit according to their preferred objective function, provides options when things don't work as expected, and helps the user to avoid the most common pitfalls.

For instance, suppose our desired objective function is indeed the negative log likelihood function. It is simple to define as a function of the shape parameter `c` and scale only, leaving the location at its default of zero.

```{code-cell} ipython3
def nllf(params):
    c, scale = params
    X = Weibull(c=c) * scale
    return -X.logpdf(data).sum()

c_, loc_, scale_ = params
nllf((c_, scale_))

> ```

Reviewer: Using params both as a 2-tuple and 3-tuple in the same codeblock is confusing IMO. I would keep the custom `nllf` as using a 3-typle, and then just discard `loc` when unpacking `params`:
```
    c, _, scale = params
```

Author: I will try to make that more clear. It will need to be just 2-tuple in the optimization, though.

Reviewer: I would still prefer that the `params` in the signature of the (here only local) `nllf` matches the 3-tuple params in the rest of the notebook, rather than introducing a 2-tuple variant just for this code-block

Author: To make sure we're talking about the same thing, it's not just in this code block: there are only two parameters being optimized, so all the code blocks that define an object function do this. 

So we would have to show `loc` being added as a decision variable and then constrained to be equal to 0. Many solvers have trouble with that. We happened to have given some `minimize` solvers the ability to handle that, but not all, so I would rather avoid show that. 

Can you think of other things that would address your concern? For example, usually in objective functions, the variable is just called `x`. I can change the name of the argument of objective functions to just `x`?

Reviewer 2: If we want to be completely unambiguous we could use `decision_variables` instead of `params`. That should make sure no one would mistakenly assume that `params` had some specific meaning other than just the variables being optimized.

Author: Hmm, but that is still a bit ambiguous because there is a distinction between the variables themselves and the value at that iteration. Maybe `values_of_decision_variables_at_this_iteration`?

Author: Actually, I had already removed three-element use of `params` before the last commit. I was just immediately unpacking calls to `weibull_min.fit` into `c_, loc_, scale_`. I went ahead and removed the remaining uses of `params`, though, for good measure.

x0 = [1, 1]  # guess to get optimization started
bounds = optimize.Bounds(lb, ub)
res_mle = optimize.minimize(nllf, x0, bounds=bounds)
res_mle
```

The value of the objective function is essentially identical, and the PDFs are indistinguishable.

```{code-cell} ipython3
import matplotlib.pyplot as plt
x = np.linspace(eps, 15, 300)

c, scale = res_mle.x
X = Weibull(c=c)*scale
plt.plot(x, X.pdf(x), '-', label='numerical optimization')

c, loc, scale = params
Y = Weibull(c=c)*scale
plt.plot(x, Y.pdf(x), '--', label='`scipy.stats.fit`')

plt.hist(data, bins=np.linspace(0, 20, 30), density=True, alpha=0.1)
plt.xlabel('x')
plt.ylabel('pdf(x)')
plt.legend()
plt.ylim(0, 0.5)
plt.show()
```

However, with this approach, it is trivial to make changes to the fitting procedure to suit our needs, enabling estimation procedures other than those provided by [`rv_continuous.fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.fit.html#scipy.stats.rv_continuous.fit). For instance, we can perform [maximum spacing estimation](https://en.wikipedia.org/wiki/Maximum_spacing_estimation) (MSE) by changing the objective function as follows.

```{code-cell} ipython3

> def nlps(params):

Reviewer: Could you add a comment what acronym `nlps` stands for?

Author: Sure. In the meantime, "negative log product spacing".

def lmoment_residual(params):
    c, scale = params
    X = Weibull(c=c) * scale
    E11 = stats.order_statistic(X, r=1, n=1).mean()
    E12, E22 = stats.order_statistic(X, r=[1, 2], n=2).mean()
    lmoments_X = [E11, 0.5*(E22 - E12)]  # the first two l-moments of the distribution
    lmoments_x = stats.lmoment(data, order=[1, 2])  # first two l-moments of the data
    return np.linalg.norm(lmoments_x - lmoments_X)  # Minimize the norm of the difference

x0 = [0.4, 3]  # This method is a bit sensitive to the initial guess
res_lmom = optimize.minimize(lmoment_residual, x0, bounds=bounds)
res_lmom.x, res_lmom.fun  # residual should be ~0
```

The plots for all three methods look similar and seem to describe the data, giving us some confidence that we have a good fit.

```{code-cell} ipython3
import matplotlib.pyplot as plt
x = np.linspace(eps, 10, 300)

for res, label in [(res_mle, "MLE"), (res_mps, "MPS"), (res_lmom, "L-moments")]:
    c, scale = res.x
    X = Weibull(c=c)*scale
    plt.plot(x, X.pdf(x), '-', label=label)

plt.hist(data, bins=np.linspace(0, 10, 30), density=True, alpha=0.1)
plt.xlabel('x')
plt.ylabel('pdf(x)')
plt.legend()
plt.ylim(0, 0.3)
plt.show()
```

> Estimation does not always involve fitting distributions to data. For instance, in gh-12134, a user needed to calculate the shape and scale parameters of the Weibull distribution to achieve a given mean and standard deviation. This fits easily into the same framework.

Reviewer: `gh-12134` does not automatically create an issue-link; we should make it clickable

Author: Sure, I forgot that one.

```{code-cell} ipython3
moments_0 = np.asarray([8, 20])  # desired mean and standard deviation
def f(params):
    c, scale = params
    X = Weibull(c=c) * scale
    moments_X = X.mean(), X.standard_deviation()
    return np.linalg.norm(moments_X - moments_0)

bounds.lb = np.asarray([0.1, 0.1])  # the Weibull distribution is problematic for very small c
res = optimize.minimize(f, x0, bounds=bounds, method='slsqp')  # easily change the minimization method
res
```

## New and Advanced Features

+++

<a id='Transformations'></a>
### Transformations

Mathematical transformations can be applied to random variables. For instance, many elementary arithmetic operations (`+`, `-`, `*`, `/`, `**`) between random variables and scalars work.

For instance, we can see that the reciprocal of a Weibull-distributed RV is distributed according to `scipy.stats.invweibull`. (An "inverse" distribution is typically the distribution underlying the reciprocal of a random variable.)

```{code-cell} ipython3
c = 10.6

X = Weibull(c=10.6)  
Y = 1 / X  # compare to `invweibull`
Y.plot()

x = np.linspace(0.8, 2.05, 300)
plt.plot(x, stats.invweibull(c=c).pdf(x), '--')

> plt.show()

Reviewer: The legends for the plots in this section are off (presumably should show both a red line and a dashed green line, rather than one blue one).

![image](https://github.com/user-attachments/assets/c13e67d9-bc60-4e3c-9115-b3a0f9cc79e0)

Author: Yeah, I'll play with it. Thanks.

```{code-cell} ipython3
Gamma = stats.make_distribution(stats.gamma)
a = 0.414

X = Gamma(a=a)  
Y = stats.log(X)  # compare to `loggamma`
Y.plot()

x = np.linspace(-17.5, 2, 300)
plt.plot(x, stats.loggamma(c=a).pdf(x), '--')
plt.show()
```

`scipy.stats.truncnorm` is the distribution underlying a truncated normal random variable. Note that the truncation transformation can be applied either before or after shifting and scaling the normally-distributed random variable, which can make it much easier to achieve the desired result than `scipy.stats.truncnorm` (which inherently shifts and scales *after* truncation).

+++

a, b = 0.1, 2

X = stats.Normal()  
Y = stats.truncate(X, a, b)  # compare to `truncnorm`
Y.plot()

x = np.linspace(a, b, 300)
plt.plot(x, stats.truncnorm(a, b).pdf(x), '--')
plt.show()

> +++

Reviewer: This should be a code cell.

Author: Weird... I prepared all this as a notebook. Maybe by accident I made the cell markdown instead of code after writing it.

+++

### Visualization

We can readily visualize functions of the distribution underlying random variables using the convenience method, `plot`.

```{code-cell} ipython3
import matplotlib.pyplot as plt
ax = X.plot()
plt.show()
```

```{code-cell} ipython3
X.plot(y='cdf')
plt.show()
```

The `plot` method is quite flexible, with a signature inspired by grammar of graphics. For instance, with the argument `x` on [-10, 10], plot the `pdf` against the `cdf`.

```{code-cell} ipython3
X.plot('cdf', 'pdf', t=('x', -10, 10))
plt.show()
```

### Order Statistics
As seen in the "Fit" section above, there is now support for distributions of [order statistics](https://en.wikipedia.org/wiki/Order_statistic) of random samples from distribution. For example, we can plot the probability density functions of the order statistics of a normal distribution with sample size 4.

```{code-cell} ipython3
n = 4
r = np.arange(1, n+1)
X = stats.Normal()
Y = stats.order_statistic(X, r=r, n=n)
Y.plot()
plt.show()
```

Compute the expected values of these order statistics:

```{code-cell} ipython3
Y.mean()
```

> ## Conclusion

Reviewer: Very nice!

+++

Although change can be uncomfortable, the new infrastructure paves the road to an improved experience for users of SciPy's probability distributions. 
It is not complete, of course! At the time this guide was written, the following features are planned for future releases:

> - Discrete distributions

Author: ```suggestion
- Additional built-in distributions (besides `Normal` and `Uniform`)
- Discrete distributions
```

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

> ```{eval-rst}

Author: ```suggestion
+++ {"tags": ["jupyterlite_sphinx_strip"]}

```{eval-rst}
```

.. jupyterlite:: ../../_contents/hypothesis_bartlett.ipynb
   :new_tab: True
```

> (rv_infrastructure)=
> # Random Variable Transition Guide

Author: ```suggestion
(rv_infrastructure)=

+++

# Random Variable Transition Guide
```

]
To address these and other shortcomings, [gh-15928](https://github.com/scipy/scipy/issues/15928) proposed a new, separate infrastructure based on the latter (random variable) approach. This transition guide documents how users of `rv_continuous` and `rv_continuous_frozen` can migrate to the new infrastructure.

*Note: The new infrastructure may not yet be as convenient for some use cases, especially fitting distribution parameters to data. Users are welcome to continue to use the old infrastructure if it suits their needs; it is not deprecated at this time.*

+++

## Basics

+++

In the new infrastructure, distributions families are classes named according to `CamelCase` conventions. They must be instantiated before use, with parameters passed as keyword-only arguments.
*Instances* of the distribution family classes can be thought of as random variables, which are commonly denoted in mathematics using capital letters.

```{code-cell} ipython3
from scipy import stats
mu, sigma = 0, 1
X = stats.Normal(mu=mu, sigma=sigma)
X
```

Once instantiated, shape parameters can be read (but not written) as attributes.

```{code-cell} ipython3
X.mu, X.sigma
```

> The documentation of [`scipy.stats.Normal`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.Normal.html#scipy.stats.Normal) contains links to detailed documentation of each of its methods. (Compare, for instance, against the documentation of [`scipy.stats.norm`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html).)[^1]

Author: ```suggestion
The documentation of [`scipy.stats.Normal`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.Normal.html#scipy.stats.Normal) contains links to detailed documentation of each of its methods. (Compare, for instance, against the documentation of [`scipy.stats.norm`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html).)
```

*Note: Although only [`Normal`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.Normal.html) and [`Uniform`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.Uniform.html) have rendered documentation and can be imported directly from `stats` as of SciPy 1.15, almost all other old distributions can be used with the new interface via [`make_distribution`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.make_distribution.html) (discussed shortly).*

Methods are then called by passing only the argument(s), if there are any, not distribution parameters.

```{code-cell} ipython3
X.mean()
```

```{code-cell} ipython3
X.pdf(x)
```

For simple calls like this (e.g. the argument is a valid float), calls to methods of the new random variables will typically be faster than comparable calls to the old distribution methods.

```{code-cell} ipython3
%timeit X.pdf(x)
```

```{code-cell} ipython3
%timeit frozen.pdf(x)
```

```{code-cell} ipython3
%timeit dist.pdf(x, loc=mu, scale=sigma)
```

> Note that the calls to `X.pdf` and `frozen.pdf` are identical, and the call to `dist.pdf` is very similar - the only difference is that the call to `dist.pdf` includes the shape parameters, whereas in the new infrastructure, shape parameters are only provided when the random variable is instantiated.

Author: 807b3cb2057b8452b8a8329ca369466cacc3f474

```{code-cell} ipython3
x = np.linspace(0, 1, 1000)
```

```{code-cell} ipython3
%timeit X.cdf(x)  # new infrastructure
```

```{code-cell} ipython3
%timeit stats.gausshyper.cdf(x, a, b, c, z)  # old infrastructure
```

Comparable performance improvements can be expected for array calculations whenever the underlying functions are computed by either numerical quadrature or inversion (root-finding).

+++

### Visualization

We can readily visualize functions of the distribution underlying random variables using the convenience method, `plot`.

```{code-cell} ipython3
import matplotlib.pyplot as plt
ax = X.plot()
plt.show()
```

```{code-cell} ipython3
X.plot(y='cdf')
plt.show()
```

> The `plot` method is quite flexible, with a signature inspired by grammar of graphics. For instance, with the argument `x` on [-10, 10], plot the `pdf` against the `cdf`.

Reviewer 3: Also possibly 
```suggestion
The `plot` method is quite flexible, with a signature inspired by grammar of graphics. For instance, with the argument `x` on `[-10, 10]`, plot the `pdf` against the `cdf`.
```

> The `plot` method is quite flexible, with a signature inspired by grammar of graphics. For instance, with the argument `x` on [-10, 10], plot the `pdf` against the `cdf`.

Author: ```suggestion
The `plot` method is quite flexible, with a signature inspired by the [Grammar of Graphics](https://books.google.com/books?id=_kRX4LoFfGQC&printsec=frontcover&source=gbs_ge_summary_r&cad=0#v=onepage&q&f=false). For instance, with the argument `x` on $[-10, 10]$, plot the PDF against the CDF.
```

> The `plot` method is quite flexible, with a signature inspired by grammar of graphics. For instance, with the argument `x` on [-10, 10], plot the `pdf` against the `cdf`.

Author: 807b3cb2057b8452b8a8329ca369466cacc3f474

#### Maximum Likelihood Estimation

The last method of the old infrastructure is `fit`, which estimates the location, scale, and shape parameters of an underlying distribution given a sample from the distribution.

```{code-cell} ipython3
c_, loc_, scale_ = stats.weibull_min.fit(data)
c_, loc_, scale_
```

The convenience of this method is a blessing and a curse. 

When it works, the simplicity is much appreciated. For some distributions, analytical expressions for the [maximum likelihood estimates](https://en.wikipedia.org/wiki/Maximum_likelihood_estimation) (MLE) of the parameters have been programmed manually, and for these distributions, the `fit` method is quite reliable. 

For most distributions, however, fitting is performed by numerical minimization of the negative log-likelihood function. This is not guaranteed to be successful - both because of inherent limitations of MLE and the state of the art of numerical optimization - and in these cases, the user of `fit` is stuck. However, a modicum of understanding goes a long way toward ensuring success, so here we present a mini-tutorial of fitting using the new infrastructure. 

> First, note that MLE is one of several potential strategies for fitting distributions to data. It does nothing more than find the values of the distribution parameters for which the negative log-likelihood (NLLF) is minimized. Note that for the original data, the NLLF is:

Author: ```suggestion
First, note that MLE is one of several potential strategies for fitting distributions to data. It does nothing more than find the values of the distribution parameters for which the negative log-likelihood (NLLF) is minimized. Note that for the distribution from which the data were generated, the NLLF is:
```

```{code-cell} ipython3
stats.weibull_min.nnlf((c, loc, scale), data)
```

The NLLF of the parameters estimated using `fit` are lower:

```{code-cell} ipython3
stats.weibull_min.nnlf((c_, loc_, scale_), data)
```

> Therefore, `fit` considers its job done, whether or not this satisfies the user's notions of a "good fit" or whether the parameters are within reasonable bounds. It gives the user no control over the process (other than specifying a guess).

Author: ```suggestion
Therefore, `fit` considers its job done, whether or not this satisfies the user's notions of a "good fit" or whether the parameters are within reasonable bounds. Besides allowing the user to specify a guess for distribution parameters, it gives little control over the process. (*Note: technically, the `optimizer` argument can be used for greater control, but this is not well documented, and its use negates any convenience offered by the `fit` method.*)
```

stats.weibull_min.nnlf(params, data)
```

Therefore, `fit` considers its job done, whether or not this satisfies the user's notions of a "good fit" or whether the parameters are within reasonable bounds. It gives the user no control over the process (other than specifying a guess).

+++

> Another problem is that MLE is an inherently problematic for some distributions. In this example, for instance, we can make the NLLF as low as we please simply by setting the location to the minimum of the data - even for bogus values of the shape and scale parameters.

Author: ```suggestion
Another problem is that MLE is inherently problematic for some distributions. In this example, for instance, we can make the NLLF as low as we please simply by setting the location to the minimum of the data - even for bogus values of the shape and scale parameters.
```

```{code-cell} ipython3
def nllf(params):
    c, scale = params
    X = Weibull(c=c) * scale
    return -X.logpdf(data).sum()

c_, loc_, scale_ = params
nllf((c_, scale_))
```

To perform the minimization, we can use `scipy.optimize.minimize`.

```{code-cell} ipython3
from scipy import optimize
eps = 1e-10  # numerical tolerance to avoid invalid parameters
lb = [eps, eps]  # lower bound on `c` and `scale`
ub = [10, 10]  # upper bound on `c` and `scale`
x0 = [1, 1]  # guess to get optimization started
bounds = optimize.Bounds(lb, ub)
res = optimize.minimize(nllf, x0, bounds=bounds)
res
```

The value of the objective function is almost identical, and the PDFs are indistinguishable.

```{code-cell} ipython3
import matplotlib.pyplot as plt
x = np.linspace(0, 15, 300)

c, scale = res.x
X = Weibull(c=c)*scale
plt.plot(x, X.pdf(x), '-', label='numerical optimization')

c, loc, scale = params
Y = Weibull(c=c)*scale
plt.plot(x, Y.pdf(x), '--', label='`scipy.stats.fit`')

plt.hist(data, bins=np.linspace(0, 20, 30), density=True, alpha=0.1)
plt.xlabel('x')
plt.ylabel('pdf(x)')
plt.legend()
```

> However, wi

Author: Hah I got distracted. But this is where I was going to show that we can easily adjust the objective function.
